server/segment_rpc_test.go

```diff
@@ -2261,6 +2261,30 @@ func TestSubmitSegment_Success(t *testing.T) {
 	balance.AssertCalled(t, "Credit", ratMatcher(change))
 }
 
+func TestSendReqWithTimeout(t *testing.T) {
+	assert := assert.New(t)
+
+	var wg sync.WaitGroup
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		wg.Wait()
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer server.Close()
+	req, _ := http.NewRequestWithContext(context.Background(), "POST", server.URL, nil)
+
+	// no timeout
+	resp, err := sendReqWithTimeout(req, 5*time.Second)
+	assert.NoError(err)
+	assert.Equal(200, resp.StatusCode)
+
+	// timeout
+	wg.Add(1)
+	resp, err = sendReqWithTimeout(req, time.Millisecond)
+	wg.Done()
+	assert.Nil(resp)
+	assert.ErrorIs(err, errTimeout)
```

Comment:
Another check to add here is whether the request continued executing on the server after the client cancellled the request. You could check that by, in the server handler, writing to some channel that you can read here. For example, write `r.Context().Err()` which would be **not nil** in case the request was really cancelled for the server or `nil` if it kept running in the end (which also means it kept running for the client of `sendReqWithTimeout`).

Reply:
I think it's not so easy to test it in a reliable manner, because `cancel()` is called by `timeouter` asynchronously (in a separate routine). That means that we can't rely on when the `r.Context().Err()` is actually set to some value.

We could do the following, but I don't like that we use this `time.sleep()` everywhere. Or maybe you have some other idea how to test it?

```golang
func TestSendReqWithTimeout(t *testing.T) {
	assert := assert.New(t)

	var wg sync.WaitGroup
	ctxErrCh := make(chan error, 1)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		wg.Wait()
		w.WriteHeader(http.StatusOK)
		time.Sleep(100 * time.Millisecond)  // to reduce the flakiness of the test
		ctxErrCh <- r.Context().Err()
	}))
	defer server.Close()
	req, _ := http.NewRequestWithContext(context.Background(), "POST", server.URL, nil)

	// no timeout
	resp, err := sendReqWithTimeout(req, 5*time.Second)
	assert.NoError(err)
	assert.Equal(200, resp.StatusCode)
	assert.NoError(<-ctxErrCh)

	// timeout
	wg.Add(1)
	resp, err = sendReqWithTimeout(req, time.Millisecond)
	wg.Done()
	assert.Nil(resp)
	assert.ErrorIs(err, context.Canceled)
	assert.ErrorIs(<-ctxErrCh, context.Canceled)
}
```

Reply:
Ah I see! I tried thinking of some way of guaranteeing the test consistency, but it does feel kind of hard given the concurrent nature of the problem here. Nevermind this, I think we can leave it as is.